_posts/2020-05-04-strangler-pattern.md

```diff
@@ -0,0 +1,107 @@
+
+Evolutionary architecture, doesn't aim for an end state.  One tool for evolutionary architecture is the [Strangler Pattern](https://martinfowler.com/bliki/StranglerFigApplication.html).
+
+# The Context
+When considering migration to the cloud, [Infrastructure Modernisation](https://www.mechanicalrock.io/our-expertise/infrastructure-modernisation), or [Serverless technology](https://www.mechanicalrock.io/our-expertise/cloud-native-solutions) how can we move faster whilst dealing with legacy systems?
+
+You can't stop everything, down tools and re-write from scratch.  What's more, if you tried you would likely be caught out by the [Second System Effect](https://wiki.c2.com/?SecondSystemEffect).  "We can solve the issues of the past and re-write the system that took us 6+ years, in 12 months."  The reality?  18 months down the line: you are still building on the original system; the new system supports ~20% of the functionality of the original and technical debt accrued in the push deliver has turned your beautiful clean new architecture into a big ball of mud.
+
+Instead, you need a mechanism to refactor and replace parts of a system, whilst maintaining production availability.
+
+# The solution
+
+The strangler pattern provides an approach to controlled re-architecture.  
+
+It is not intended persistent indefinitely.  Instead it provides a pathway to manage internal change whilst minimising the impact to external dependencies.  Once re-architecture has been completed, a key step is the removal of the strangler from your system to simplify and move forward.
+
+![Strangler Pattern](/img/strangler-pattern/strangler-pattern.png)
+
+**Stage 1**: Identify the system boundary you want to re-architect.
+
+**Stage 2**: Create a [facade](https://en.wikipedia.org/wiki/Facade_pattern).
+
+* The facade ensures clients of the system are unaffected by the re-architecture activity, and should therefore match an existing system API.
+* The API for the new service should be carefully designed to ensure backward compatability with the legacy API.  If you want to make breaking changes to the API for the new service, that should be performed once re-architecture is completed.
+
+**Stage 3**: Re-architect.
+
+* Implement your new service with a compatible.
+* Dark launch the new service: to support a controlled rollout, and tee traffic between the new and legacy implementations.
+* Return responses from the legacy implementation.  
+* Capture, or discard, responses from the new implementation.  This enables you to monitor the behaviour of the new implementation to ensure it is fit for purpose.
+* This also provides a mechanism to support continuous deployment: during development, deploy partial implementations of the new service, to gain confidence and mitigate risk without compromising system behaviour
```

Comment:
This last point reads a bit funny. Maybe use a full stop instead of a colon and reign the commas in?